### PR TITLE
feat: upgrades for verifier crate

### PR DESCRIPTION
**Pull request by yuwen01**
1. Serialize and deserialize Fq2 in the same format as gnark and arkworks.
2. Flag SP1 inclusion under cfg flag.
3. Refactor inverse_unconstrained and inverse to inverse and cpu_inverse. This makes it easier to choose the correct implementation depending on architecture.
4. Corrected version from 0.7.0 to 0.6.0.

I want to use this branch as the v2 feature branch for now, then merge it into patch-v0.6.0, which I imagine to be our main branch. 

| Operation | V1 Cycle count (mil) | V2 Cycle Count (mil) |
|--------|--------|--------|
| 2x Batch Pairing | 6.09  | 5.176 |
| Groth16 Verification | unimplemented | 6.77 |

### CODE REVIEW

File: src/fields/fq2.rs [hunk 1]
```diff
@@ -255,15 +256,14 @@ impl FieldElement for Fq2 {
...
-            .cpu_sub((self.c1.cpu_mul(self.c1)).cpu_mul(fq_non_residue())))
-        .inverse()
-        .map(|t| Fq2 {
-            c0: self.c0.cpu_mul(t),
-            c1: (self.c1.cpu_mul(t)).cpu_neg(),
-        })
+        self.c0
+            .mul(self.c0)
+            .sub((self.c1.mul(self.c1)).mul(fq_non_residue()))
+            .inverse_unconstrained()
```
Comment (ratankaliani) [inline src/fields/fq2.rs]: Should we gate this behind a `zkvm` feature flag given that we're changing it from `cpu_mul` to zkvm `mul`, and `inverse_unconstrained`

File: src/fields/fq2.rs [hunk 2]
```diff
@@ -255,15 +256,14 @@ impl FieldElement for Fq2 {
...
+            .map(|t| Fq2 {
+                c0: self.c0.cpu_mul(t),
+                c1: (self.c1.cpu_mul(t)).cpu_neg(),
+            })
```
Comment (ratankaliani) [inline src/fields/fq2.rs]: And why are we using `cpu_mul` here?

It'd be great to include a comment for why we're using different functions from the orig

File: src/groups/mod.rs [hunk 3]
```diff
@@ -347,7 +348,7 @@ impl<P: GroupParams> G<P> {
                 y: self.y,
             })
         } else {
-            let zinv = self.z.inverse_unconstrained().unwrap();
+            let zinv = self.z.inverse().unwrap();
```
Comment (ratankaliani) [inline src/groups/mod.rs:351]: Why did you change this from `inverse_unconstrained`

File: src/fields/fq2.rs (hunk 1 above)
Comment (yuwen01) [reply src/fields/fq2.rs]: cpu_mul -> mul
This was causing the cycle count to go up by like 10% per pairing. mul and sub already check which environment we're running the patch in before correctly routing to either cpu_mul or the syscall mul -- there's no reason to force cpu_mul at this level.

also inverse_unconstrained automatically routes to inverse if the target os is not zkvm

File: src/fields/fq2.rs (hunk 2 above)
Comment (yuwen01) [reply src/fields/fq2.rs]: u right these can be mul and neg

basically there are redundant checks for which os we're in. 

In general, I think differentiating between inverse and inverse unconstrained is a bad idea. The FieldElement trait should just expose inverse, and it should compile to the specialized unconstrained implementation or the default implementation depending on the target os. Simplifying the interface in this way will make errors like this impossible, and will probably reduce cycle count too.

**Review by ratankaliani — COMMENTED** [review]
Given that we feature flag inside of `mul` and `sub`, we should use these inside of the patch.

Find the bug in `inverse_unconstrained`, and then modify `inverse` to feature flag between the original implementation and `inverse_unconstrained`.

File: src/fields/fp.rs [hunk 4]
```diff
@@ -130,12 +130,6 @@ impl Fr {
         ])
     }
 
-    #[inline]
-    #[allow(dead_code)]
```
Comment (yuwen01) [inline src/fields/fp.rs:134]: I have no idea why this was here in the first place.

File: src/fields/fq2.rs [hunk 5]
```diff
@@ -255,25 +255,13 @@ impl FieldElement for Fq2 {
...
-    fn inverse_unconstrained(self) -> Option<Self> {
         #[cfg(target_os = "zkvm")]
         {
             // Compute the inverse using the zkvm syscall
             sp1_lib::unconstrained! {
                 let mut buf = [0u8; 65];
-                self.inverse().map(|inv| {
-                    let bytes = cast::<[u128; 4], [u8; 64]>(inv.to_u512().0);
+                self.cpu_inverse().map(|inv| {
+                    let bytes = cast::<Fq2, [u8; 64]>(inv);
```
Comment (yuwen01) [inline src/fields/fq2.rs:264]: This serialization was the inverse bug

File: src/groups/mod.rs (hunk 3 above)
Comment (yuwen01) [reply src/groups/mod.rs:351]: inverse_unconstrained is now deprecated
